I-Soil Verification Documentation for SMiRT #234

crswong888 opened the pull request:
Created V&V documentation for I-Soil backbone curve types and pressure dependency parameters.
- Restructured the `doc/content/sqa` directory and created a directory to store the markdown files for the I-Soil V&V page.
- Created a directory `doc/content/media/materials/isoil` to store images used in the I-Soil verification page.
- Added files to the `gold` directory under `isoil` to be used in the documentation.
- Corrected the `HYS_pressure_dependent_stiffness.i` test - previously, it was actually verifying pressure dependent strength.
- Created a new test: `HYS_pressure_dependent_strength.i` and added it to the `tests` file as `RunException`.

Some charts were added as `png` images since they could not be plotted using MooseDown plotly. I have opened an [issue](https://github.com/idaholab/moose/issues/13855) in idaholab/moose regarding this.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/377354/) on 6d8d8b7 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/mastodon/docs/PRs/234/site)

This comment will be updated on new commits.

cbolisetti commented on doc/content/sqa/v_and_v/materials/isoil/pressure_dependency.md:
```diff
@@ -0,0 +1,139 @@
+
+!listing test/tests/materials/isoil/HYS_stiffness_and_strength_pressure_dependency.i
+
+*Some modifications to this input file were made to produce the results presented on this page (i.e., the different parameter values used for the two cases). Please refer to [this section](#strength-stiffness) for a description of these changes.*
```
Same

cbolisetti commented on doc/content/sqa/v_and_v/materials/isoil/problem_statement.md:
```diff
@@ -0,0 +1,20 @@
+## Problem Statement id=problem-statement
+
+A single hexagonal element with dimensions of 1 meter *x* 1 meter *x* 1 meter was used for all verifications presented in this report. The back surface ($z=0$) of the element was fixed in all three principal directions and the ramp displacement, $u_{g}$, shown in [applied-displacement] was applied to the front surface ($z=1$) in the X direction.
```
Hexahedral, not hexagonal. Please change this in all occurrences.

cbolisetti commented on doc/content/sqa/v_and_v/materials/isoil/user_defined.md:
```diff
@@ -0,0 +1,40 @@
+
+!listing test/tests/materials/isoil/HYS_GQH.i
+
+*Some modifications to this input file were made to produce the results presented on this page (e.g., initial stress and gravity were omitted). Please refer to [this section](#user-defined) for a description of these changes.*
```
Same.

cbolisetti commented on doc/content/sqa/v_and_v/materials/isoil/pressure_dependency.md:
```diff
@@ -0,0 +1,139 @@
+
+The three cases described in [stiffness-cases] were used to investigate the behavior of [p-ref-modulus] with different values for the parameter $b$. In all cases, the pressure dependent strength parameters were set to the following: $a_{0}=1$, $a_{1}=0$, and $a_{2}=0$, therefore, the yield strength was not modified, and $p_{ref}$ was set to 6072.86 Pa.
+
+*+Warning+*: If all three pressure dependent dependent strength parameters are set to zero in the input file, the solver will struggle to converge. One can simply set $a_{0}=1$ and the others to zero if they truly want to ignore shear strength dependency.
```
If all three are zero, we need to throw an error, if we aren't throwing one already. Please add that to the code, if you can -- both in ComputeISoilStress.C and ISoilAction.C

crswong888 commented on doc/content/sqa/v_and_v/materials/isoil/gqh.md:
```diff
@@ -0,0 +1,83 @@
+
+!listing test/tests/materials/isoil/HYS_GQH.i
+
+*Some modifications to this input file were made to produce the results presented on this page (e.g., initial stress and gravity were omitted). Please refer [this section](#gqh) for a description of these changes.*
```
No problem, I will make the input file shown here the exact one which produces these results. Would you suggest that I modify the existing test input file? Or create an entirely new test. The inputs are very similar, so modifications would be minor. 

Although, those modifications highly simplify the problem, perhaps the existing tests are intentionally complex?

cbolisetti commented:
I think redundancy in tests is not a bad thing. Please create new tests and
provide appropriate requirements for them.

The existing tests weren’t intended for verification — just testing. So we
didn’t think about their complexity at the time.

@sveerara, @aeslaughter, please let me know if you disagree.

On Thu, Aug 8, 2019 at 11:42 AM crswong888 <notifications@github.com> wrote:

> *@crswong888* commented on this pull request.
> ------------------------------
>
> In doc/content/sqa/v_and_v/materials/isoil/gqh.md
> <https://github.com/idaholab/mastodon/pull/234#discussion_r312107537>:
>
> > +                      'title':'Hysteresis Loop in the ZX Plane'}
> +              id=gqh-hysteresis-loop
> +              caption=Plot of the element hysteresis stress-strain curve in the ZX plane over the duration of the simulation.
> +
> +Since the engineering strain in the ZX plane, $\gamma_{zx}$, is equal to two times the tensor shear strain, $\epsilon_{zx}$, it was necessary to multiply the strains output by MOOSE TensorMechanics by two to get the shear stresses, $\tau_{zx}$, which correspond to those determined by [backbone]. Next, the engineering shear stress-strain curve was superimposed over the 100 points, calculated by hand with [curve] and [backbone], as shown in [gqh-backbone-curve]. The backbone curve was obtained in MOOSE by sampling the data during the initial load cycle, i.e., from 3 seconds to 4 seconds (see [#problem-statement]). It is clear from [gqh-backbone-curve] that the analytical solution of the GQ/H model was consistent with the MASTODON one.
> +
> +!media media/materials/isoil/GQH_backbone.png
> +       id=gqh-backbone-curve
> +       caption=The analytical GQH backbone curve sampled at 100 points and the engineering shear stress-strain curve in the ZX plane output by MASTODON during the initial loading cycle (from 3 seconds to 4 seconds).
> +       style=width:100%
> +
> +### Complete Input File
> +
> +!listing test/tests/materials/isoil/HYS_GQH.i
> +
> +*Some modifications to this input file were made to produce the results presented on this page (e.g., initial stress and gravity were omitted). Please refer [this section](#gqh) for a description of these changes.*
>
> No problem, I will make the input file shown here the exact one which
> produces these results. Would you suggest that I modify the existing test
> input file? Or create an entirely new test. The inputs are very similar, so
> modifications would be minor.
>
> Although, those modifications highly simplify the problem, perhaps the
> existing tests are intentionally complex?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/mastodon/pull/234?email_source=notifications&email_token=ACQ6BUZQVNXTUT3QZ7VFQKTQDQ5HXA5CNFSM4IJBSMHKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBAIR7Q#discussion_r312107537>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQ6BU4DT3TSPZSRCX35CKDQDQ5HXANCNFSM4IJBSMHA>
> .
>
-- 
Sent from Mobile

crswong888 commented:
Okay no problem, I'll create the new tests as Exodiffs. What should the new nomenclature for these tests be? I can name the modified input files the same as the originals, followed by a suffix such as, "`_simple.i`", e.g., the modified test for `HYS_darendeli.i` could be named `HYS_darendeli_simple.i`. Or perhaps there's a more appropriate suffix than "simple", ... or a prefix.

cbolisetti commented:
Add verification as a suffix

On Thu, Aug 8, 2019 at 12:02 PM crswong888 <notifications@github.com> wrote:

> Okay no problem, I'll create the new tests as Exodiffs. What should the
> new nomenclature for these tests be? I can name the modified input files
> the same as the originals, followed by a suffix such as, "_simple.i",
> e.g., the modified test for HYS_darendeli.i could be named
> HYS_darendeli_simple.i. Or perhaps there's a more appropriate suffix than
> "simple", ... or a prefix.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/mastodon/pull/234?email_source=notifications&email_token=ACQ6BU2WVZ6KNNUPOETCGJTQDQ7QJA5CNFSM4IJBSMHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD34C7EI#issuecomment-519581585>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQ6BU2KMVV45BJHFIKR7DTQDQ7QJANCNFSM4IJBSMHA>
> .
>
-- 
Sent from Mobile

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/379461/) on b63a801 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/mastodon/docs/PRs/234/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/mastodon/docs/PRs/234/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7300db9aebeb9d1ec15dcf99a2f6aaa0fb07f334`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/379464/) on 501b2f9 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/mastodon/docs/PRs/234/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/mastodon/docs/PRs/234/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7300db9aebeb9d1ec15dcf99a2f6aaa0fb07f334`

crswong888 commented:
The mooseError I just added works for the I-Soil Action System as well, since `ISoilAction.C` uses `InputParameters params = _factory.getValidParams("ComputeISoilStress");`

crswong888 commented:
I see that this has not been merged yet, may I commit a couple very small changes? Some of the `#` comments I put in the new test files I created are wrong information.

cbolisetti commented:
Yes, please do! Sorry I didn’t get a chance to look at it again.

On Fri, Aug 9, 2019 at 12:21 PM crswong888 <notifications@github.com> wrote:

> I see that this has not been merged yet, may I commit a couple very small
> changes? Some of the # comments I put in the new test files I created are
> wrong information.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/mastodon/pull/234?email_source=notifications&email_token=ACQ6BU67GWWI2HHWI6LEALTQDWKSHA5CNFSM4IJBSMHKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD37EIHI#issuecomment-519980061>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQ6BU7AYWOG2UCLWHOKFSDQDWKSHANCNFSM4IJBSMHA>
> .
>
-- 
Sent from Mobile

cbolisetti commented:
@crswong888 these are a lot of commits. Can you please squash them into 1-2 commits. You can find directions on how to do this online.

crswong888 commented:
So I have squashed all of the commits labeled #234, which are the ones that I actually made, into just two commits: [`88041f0`](https://github.com/idaholab/mastodon/pull/234/commits/88041f018788b5ff38159da9b5e277110cbbaa8f) and [`4690b64`](https://github.com/idaholab/mastodon/pull/234/commits/4690b643d22e0860827ce6dcbae6484e5a52589f). But it looks like in my various cherry-picking and rebasing procedures that I've added myself as a contributor to all of the commits that were simply merged from `idaholab/mastodon` into my fork after I opened this PR.

@cbolisetti, what do you I suggest I do here? The most recent two commits are all of the work I've done for this PR. I supposed I could just cherry-pick my two commits and drop all of the merges, assuming that won't create any conflicts. Also, the apparent moose conflict is fixed, I just forgot to update the submodule, so it should be fixed next time I push.

crswong888 commented:
@cbolisetti Nevermind, I just did a hard reset with `idaholab/mastodon` and cherry-picked my commits. There is now just 2 commits for this PR. First, one for all the I-Soil verification files, including markdowns, media, `csv` files for plots, and the new tests. And a second one for the `MooseError` I added to `ComputeISoilStress.C`. Let me know if there is anything else you need for this PR.

crswong888 commented:
In [this third commit](https://github.com/idaholab/mastodon/pull/234/commits/2be13949c79a938d3b5fdb34ca6844bc7d8e7202), I added the phrase `"...If 'soil_type = darendeli', then the reference pressure must be input in kilopascals."` to the `ValidParams` description of `p_ref` in `ComputeISoilStress.C`, `ISoilAction.C`, and `ADComputeISoilStress.C`. This will, at least, make the user immediately aware of this requirement for the MASTODON implementation of the Darendeli model until a more robust resolution to [this issue](https://github.com/idaholab/mastodon/issues/216) is implemented.

crswong888 commented:
![Screenshot from 2019-08-20 20-16-02](https://user-images.githubusercontent.com/41302548/63397352-503b0000-c398-11e9-90d8-50ee4d1087d4.png)

cbolisetti commented on doc/content/sqa/v_and_v/materials/isoil/pressure_dependency.md:
```diff
@@ -0,0 +1,131 @@
+       id=strength-comparison2
+       caption=The user-defined backbone curve and the engineering shear stress-strain curve in the ZX plane output by MASTODON during the initial loading cycle (from 3 seconds to 4 seconds) for pressure dependent strength Cases 2, 4, and 5.
+
+Since setting $a_{2}=1$ and $a_{1}=0$ causes the shear strength to increase exponentially, while setting $a_{2}=0$ and $a_{1}=1$ causes it to increase only linearly, [strength-comparison1] shows that Case 1 is significantly stronger than Case 2. Also, since setting $a_{0}=1$ while setting the others to zero causes the right-hand-side of [p-ref-strength] to be equal to unity, [strength-comparison1 ...
```
'Since setting $a_{2}=1$ and $a_{1}=0$ causes the shear strength to increase exponentially, while setting $a_{2}=0$ and $a_{1}=1$ causes it to increase only linearly,' 

This is unclear/inaccurate -- case 1 (a2=1 and others zero) makes the strength to increase linearly with pressure since there is a square root. In case 2, it is proportional to the sq root of the pressure and in case 3 it is independent of the pressure. Please recheck the sentence,.

cbolisetti commented on test/tests/kernels/time_integration/tests:
```diff
@@ -3,7 +3,7 @@
     input = 'newmark.i'
     exodiff = 'newmark_out.e'
-    verification = 'V1-01.md'
+    verification = 'newmark.md'
```
In this file, can you please change the word validation to verification?

cbolisetti commented on test/tests/kernels/time_integration/tests:
```diff
@@ -3,7 +3,7 @@
-    verification = 'V1-01.md'
+    verification = 'newmark.md'
     requirement = 'The Newmark time integration scheme in MASTODON shall calculate results that are equal to analytical calculations.'
     design = 'timeintegrators/NewmarkBeta.md'
```
also in this file.

cbolisetti commented on test/tests/kernels/time_integration/tests:
```diff
@@ -12,7 +12,7 @@
     input = 'hht.i'
     exodiff = 'hht_out.e'
-    verification = 'V1-02.md'
+    verification = 'hht.md'
```
also here. The comparison done here is actually verification, not validation, as described in the .md file. I know you didnt write it, but it would be helpful if you could just change it in this PR.

cbolisetti commented on test/tests/materials/isoil/tests:
```diff
@@ -232,16 +265,28 @@
     expect_err = "Pressure dependency is set to true but b\_exp is set to 0\.0\. Stiffness pressure dependency is NOT turned on\."
 
-    requirement = "The ComputeISoilStress class shall generate a warning if pressure dependency is set to true but b\_exp is set to 0.0."
+    requirement = "The ISoilAction and ComputeISoilStress class shall accurately adjust the stress-strain curve to account for pressure dependent changes in strength of the soil, but not changes in stiffness."
```
Please remove, 'but not changes in stiffess'. the stiffness (at the same strain) will change as well, when the strength is increased.

crswong888 commented:
Okay @cbolisetti, I implemented all the changes you requested except the two regarding the pressure dependent strength stuff. I agree, my wording here was technically incorrect. Perhaps there is some way I can write a more precise statement. 

I took a moment to make the following plot (which, interestingly, is sort of a visual representation of what the predictor-corrector algorithm is doing):

![backbone_interpolation](https://user-images.githubusercontent.com/41302548/64480425-affd1d80-d184-11e9-85a4-5c5a1d41b978.png)

This graph shows the output for Case 3 from the strain of 0.01 to 0.02. All I did here was interpolate out from each stress using the tangent modulus defined by the original backbone curve. It can be assured that wherever the solid line diverges from the dashed line, that the element is yielding, and therefore the tangent stiffness is indirectly modified as you are suggesting. However, my original thought when I wrote this was that MASTODON is not *directly* modifying the tangent stiffness, i.e., its not using the equation which controls pressure dependent stiffness in any way. The strain increments shown on the plot that match prove this. In the same way, for exclusively pressure dependent stiffness, the strength for a given strain increment is indirectly modified. 

I'd hate to make this PR any more complicated. But I would be happy to include a plot like the one I showed here in the actual documentation and write a more precise statement about what is going on. Or, I could simply remove the misleading sentences and submit this PR as it is. Let me know what you would like me to do.

cbolisetti commented:
@crswong888 I see what you mean, but the little kinks in that plot dont really affect the global response. Therefore, I think it would be a little confusing for the users if we say that the stiffness doesnt change when we use strength dependency only. So can you please update the text and the requirements accordingly? 

On another note, @sveerara and I were looking at the above plot and were a little concerned with the sudden changes in stress. Is the plot you showed stress vs strain for case 3? We are unable reproduce this exact plot, for some reason.

crswong888 commented:
Okay, @cbolisetti, I made some changes to the pressure dependency section of the report and changed the `requirements` wording for the respective tests. Since I presented this report in such a way that the equations were basically the principle argument, and then I subsequently split up the section to individually examine the two equations, `[p-ref-strength]` and `[p-ref-modulus]`, as I have named them, I felt the need to say something, like: "well, technically, the stiffness and strength weren't directly modified, but it's quite impossible to change one and not the other holding all else constant." Hopefully my edits are clear. I'll squash my last two commits once you get a chance to take a look at [this most recent one](https://github.com/idaholab/mastodon/pull/234/commits/c12715ca5fd1c5d588be83205c8472fc1b8580c2).

Regarding yours and @sveerara attempt at reproducing my plot, this is my mistake!! It is actually *Case 1*, not Case 3 as I mistakenly said. This, I'm sure, probably makes a difference. But just to be clear: I don't believe there's any need to be concerned, the jagged black line showing the stepping pattern were my own calculations, completely separate from MOOSE, in a spreadsheet. The dashed line is the output from MOOSE.

If you guys' are still interested, below is a description of my procedure

### How I obtained that plot: ###
- The backbone curve defined for this model, which already existed in the `test` directory when I was working on these files this summer, was the following:

| stress  | strain |
| ------------- | ------------- |
| 1e-06 | 20.2074269855 |
| 3e-06 | 59.5750529544 |
| 1e-05 | 187.3933535273 |
| 3e-05 | 486.5032012444 |
| 0.0001 | 1134.1888160611 |
| 0.0003 | 1986.3342141321 |
| 0.001 | 3184.7207919449 |
| 0.003 | 4528.9169131199 |
| 0.007 | 5569.4392947359 |
| 0.01 | 5955.680180984 |

- my intent was to examine, more closely, how the stiffness resulting from the strength modification differed from the the tangent stiffness defined by the original backbone curve. So I picked the ordinates for which the element was strained beyond the backbone input, because my assumption was that MASTODON will continue follow the most recently defined slope, which I am pretty sure is true based on my best effort interpretation of `ISoilUtils.C`.
- The time interval I chose went precisely from time step 352 to 403, where the strain ranges from 0.01 to 0.0197:

![Screenshot from 2019-09-19 21-19-36](https://user-images.githubusercontent.com/41302548/65296922-61a43300-db23-11e9-84c4-fa89734a6714.png)

- The plot I showed [here](https://user-images.githubusercontent.com/41302548/64480425-affd1d80-d184-11e9-85a4-5c5a1d41b978.png) was basically a zoom-in of the region in red. Thus, the tangent stiffness in this interval should, I believe, follow the final slope available from the backbone curve, i.e.,

![image](https://user-images.githubusercontent.com/41302548/65296758-bf844b00-db22-11e9-9be9-87210c259499.png)

- To compare, I copied the engineering stress-strain (engineering strain is twice the shear strain) `.csv` data from moose and extrapolated out from each stress over the same strain increment using the _k_ I derived. Then I forcefully shifted my data back up to match the MOOSE data at the next strain point. So, in general, the curve followed this series of points: 

1. moose data at point _i_ 
2. extrapolate out to point _i_ + 1 using assumed _k_
3. moose data at point _i_ + 1, and so on.

- Thus, I think the "kinks" that you are referring to are no cause for alarm, since they were entirely fabricated by myself.
- Again, I merely wanted compare the stiffness resulting from the strength modification compared with that of the original backbone curve. In my plot, I had shown that, although the stiffness has apparently also been modified, even though I set `b=0` in the input file, it is not being _directly_ modified by MASTODON, which is proven by the strain increments for which the output follows a _completely separate_ extrapolation I did _by hand_.

One final point, I think it only makes sense that this is the case, since, if we take a look at `ComputeISoilStress.C`:

```
  if (_pressure_dependency)
  {
    _stiffness_pressure_correction = pow((mean_stress - _p0) / _p_ref[_pos], _b_exp);
    _strength_pressure_correction =
        std::sqrt(_a0 + _a1 * (mean_stress - _p0) +
                  _a2 * (mean_stress - _p0) * (mean_stress - _p0)) /
        std::sqrt(_a0 + _a1 * (_p_ref[_pos]) + _a2 * (_p_ref[_pos]) * (_p_ref[_pos]));
  }
```
where, clearly, `_stiffness_pressure_correction = 1` if `_b_exp = 0`. Then we come down here:
```
    // compute the deviatoric trial stress normalized by non pressure dependent
    // youngs modulus.
    _deviatoric_trial_stress =
        _individual_stress_increment.deviatoric() * _stiffness_pressure_correction +
        (*_stress_model_old[i])[_qp] / (_youngs[_pos][i]);

    // compute the effective trial stress
    Real dev_trial_stress_squared =
        _deviatoric_trial_stress.doubleContraction(_deviatoric_trial_stress);
    Real effective_trial_stress = std::sqrt(3.0 / 2.0 * dev_trial_stress_squared);

    // check yield condition and calculate plastic strain
    Real yield_condition =
        effective_trial_stress - _yield_stress[_pos][i] * _strength_pressure_correction;
```
In this code, the `_deviatoric_trial_stress` should be what it would otherwise be as determined from `ISoilUtils.C`. Therefore, only the `_strength_pressure_correction` should effect the `yield_condition`. I think, at least. I'm just trying to make sense of this beautiful material model!

cbolisetti commented:
Thanks a lot for that explanation @crswong888. I still need to look into the last part about the yield condition. I am merging this commit for now. This is good work, and we can make changes in the future if needed.

crswong888 commented:
@cbolisetti Great! Sorry to drag this PR out to infinity, but I think all our scrutiny made this an accurate and informative article, and I learned a lot. I-soil is an excellent material model.

 Let me know if you need some assistance on any future projects. I'm happy to volunteer some man-hours when I can!
